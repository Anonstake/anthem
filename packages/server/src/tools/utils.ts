import bech32 from "bech32";
import moment from "moment-timezone";
import { identity, ifElse, lensProp, over } from "ramda";
import { IMsgDelegate, ITransaction, ITxMsg } from "../schema/graphql-types";
import NETWORKS, {
  NETWORK_NAME,
  NetworkDefinition,
} from "../server/sources/networks";

/** ===========================================================================
 * Types & Config
 * ============================================================================
 */

const getCosmosSdkAddressEnumsForNetwork = (name: NETWORK_NAME) => {
  const networkPrefix = name.toLowerCase();
  return {
    ACCOUNT_ADDRESS: `${networkPrefix}`,
    ACCOUNT_PUBLIC_KEY: `${networkPrefix}pub`,
    VALIDATOR_OPERATOR_ADDRESS: `${networkPrefix}valoper`,
    VALIDATOR_CONSENSUS_ADDRESS: `${networkPrefix}valcons`,
    VALIDATOR_CONSENSUS_PUBLIC_KEY: `${networkPrefix}valconspub`,
    VALIDATOR_OPERATOR_PUBLIC_KEY: `${networkPrefix}valoperpub`,
  };
};

/** ===========================================================================
 * Utils
 * ============================================================================
 */

/**
 * Assert a condition cannot occur. Used for writing exhaustive switch
 * blocks (e.g. see unwrapOkValueIfExists).
 */
export const assertUnreachable = (x: never): never => {
  throw new Error(
    `Panicked! Received a value which should not exist: ${JSON.stringify(x)}`,
  );
};

/**
 * Sorting method for transactions list result, which sorts by transaction
 * block height.
 */
export const orderTxsByHeight = (txA: ITransaction, txB: ITransaction) => {
  return Number(txA.height) > Number(txB.height) ? -1 : 1;
};

// Associate amount value to amounts key
const updateFn = over(lensProp("value"), (msg: IMsgDelegate) => ({
  ...msg,
  amounts: msg.amount,
}));

// Update amount key if it is an array
const updateAmounts = ifElse(
  ({ value }: ITxMsg) => {
    const { amount } = value as IMsgDelegate;
    return Boolean(amount && Array.isArray(amount));
  },
  updateFn,
  identity,
);

/**
 * Format the transaction response data from Clickhouse.
 */
export const formatTransactionResponse = ({
  hash,
  height,
  gaswanted,
  gasused,
  log,
  memo,
  fees,
  tags,
  msgs,
  timestamp,
  chain,
}: any): ITransaction => ({
  hash,
  height,
  timestamp,
  gaswanted,
  gasused,
  memo,
  chain,
  fees,
  tags,
  msgs: msgs.map(updateAmounts),
  log: Array.isArray(log) ? log : [log],
});

/**
 * Decode a validator address using bech32 and re-encode it to derive the
 * associated validator address.
 */
export const getValidatorAddressFromDelegatorAddress = (
  address: string,
  network: NETWORK_NAME,
): string | null => {
  try {
    const networkAddressEnum = getCosmosSdkAddressEnumsForNetwork(network);
    const decodedAddress = bech32.decode(address);
    const validatorAddress = bech32.encode(
      networkAddressEnum.VALIDATOR_OPERATOR_ADDRESS,
      decodedAddress.words,
    );

    return validatorAddress;
  } catch (err) {
    return null;
  }
};

/**
 * Remove sanity check heights from the results, all the sanity check
 * heights are at 10_000 height intervals.
 */
export const filterSanityCheckHeights = (response: { height: number }) => {
  return response.height % 10000 !== 0;
};

// TODO: Types?
export const mapSumToBalance = (item: any): any => {
  return { ...item, balance: item.sum };
};

/**
 * Verify an object contains all of the given keys.
 */
export const objectHasKeys = (
  obj: any,
  keys: ReadonlyArray<string>,
): boolean => {
  for (const key of keys) {
    if (!(key in obj)) {
      return false;
    }
  }
  return true;
};

/**
 * Determine the network for a given address using the address prefix.
 */
export const deriveNetworkFromAddress = (
  address: string,
): NetworkDefinition => {
  if (address.substring(0, 6) === "cosmos") {
    return NETWORKS.COSMOS;
  } else if (address.substring(0, 5) === "terra") {
    return NETWORKS.TERRA;
  } else if (address.substring(0, 4) === "kava") {
    return NETWORKS.KAVA;
  } else if (address.length === 44) {
    return NETWORKS.OASIS;
  }

  throw new Error(
    `Unrecognized address ${address} with no associated network!`,
  );
};

/**
 * Get the network definition from a provided network name.
 */
export const getNetworkDefinitionFromIdentifier = (networkName: string) => {
  const name = networkName.toUpperCase();

  if (name in NETWORKS) {
    return NETWORKS[name];
  }

  throw new Error(
    `Unsupported or invalid network name ${networkName} supplied!`,
  );
};

export const convertTimestampToUTC = (timestamp: string | number | Date) => {
  const UTC_ISO = moment(new Date(timestamp))
    .tz("Etc/UTC")
    .toISOString();

  return UTC_ISO;
};

// Convert all timestamps to the same ISO string format.
export const standardizeTimestamps = (item: any) => {
  return item
    .filter((x: any) => Boolean(x.timestamp))
    .map((x: any) => {
      // Force timestamp to UTC and convert to string
      const ISO = convertTimestampToUTC(new Date(x.timestamp));
      return {
        ...x,
        timestamp: ISO,
      };
    });
};
