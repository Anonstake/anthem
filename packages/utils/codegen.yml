overwrite: true
schema: ../server/src/schema.graphql
documents: schema_documents/**/*.gql
generates:
  src/graphql/introspection-result.json:
    plugins:
      - fragment-matcher
  src/graphql/types.tsx:
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-apollo
    config:
      avoidOptionals: true
      typesPrefix: I
      declarationKind: interface
      withHOC: true
      withHooks: true
      withComponent: true
      withMutationFn: true
