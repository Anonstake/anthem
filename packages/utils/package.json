{
  "name": "@anthem/utils",
  "version": "0.0.1",
  "description": "Anthem Utils",
  "private": true,
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "prettier": "prettier --list-different \"src/**/*.ts\" \"src/**/*.{ts,json}\"",
    "prettier:fix": "prettier --write \"src/**/*.ts\" \"src/**/*.{ts,json}\"",
    "tslint": "tslint --project tsconfig.json -c tslint.json",
    "tslint:fix": "tslint --project tsconfig.json -c tslint.json --fix",
    "format": "yarn prettier:fix && yarn tslint:fix",
    "lint": "yarn prettier && yarn tslint",
    "test:unit": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test": "yarn lint && yarn test:unit",
    "types": "graphql-build --outputFile=./src/schema",
    "gql": "yarn gqlgen && chmod +x scripts/gql.sh && scripts/gql.sh",
    "gqlgen": "gql-lib --schemaFilePath ../server/src/schema/schema.graphql --destDirPath ./src/client/gql --depthLimit 10",
    "record": "yarn gql && yarn mocks:refresh && yarn mocks:generate && yarn format",
    "mocks:refresh": "rm -rf src/client/data && mkdir src/client/data",
    "mocks:generate": "ts-node src/client/query-script.ts"
  },
  "dependencies": {
    "@types/bytes": "^3.1.0",
    "@types/debug": "^4.1.5",
    "bytes": "^3.1.0",
    "graphql-request": "^1.8.2",
    "prettier": "^1.19.1",
    "ts-node": "^8.8.2",
    "tslint": "^5.20.1",
    "typescript": "^3.7.4",
    "utils": "^0.3.1"
  },
  "devDependencies": {
    "@apollo/react-common": "^3.1.4",
    "@apollo/react-components": "^3.1.5",
    "@apollo/react-hoc": "^3.1.5",
    "@apollo/react-hooks": "^3.1.5",
    "gql-generator": "^1.0.12",
    "jest": "24.7.1",
    "react-graphql-generator-helper-lib": "^0.0.34"
  },
  "main": "build/main.js",
  "typings": "build/main.d.ts",
  "files": [
    "build"
  ],
  "jest": {
    "testEnvironment": "node",
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".test.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    }
  }
}
