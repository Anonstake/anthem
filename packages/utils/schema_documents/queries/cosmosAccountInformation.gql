query cosmosAccountInformation($address: String!){
    cosmosAccountInformation(address: $address){
        type
        value{
            account_number
            address
            coins{
                denom
                amount
            }
            public_key{
                type
            }
            sequence
        }
    }
}