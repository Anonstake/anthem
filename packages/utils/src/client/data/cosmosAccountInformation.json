{
  "cosmosAccountInformation": {
    "type": "cosmos-sdk/Account",
    "value": {
      "account_number": "16451",
      "address": "cosmos15urq2dtp9qce4fyc85m6upwm9xul3049um7trd",
      "coins": [{ "denom": "uatom", "amount": "10437483575" }],
      "public_key": { "type": "tendermint/PubKeySecp256k1" },
      "sequence": "78"
    }
  }
}
