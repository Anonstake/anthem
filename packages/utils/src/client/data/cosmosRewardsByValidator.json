{
  "cosmosRewardsByValidator": [
    {
      "reward": [{ "denom": "uatom", "amount": "2491958.851869806299614844" }],
      "validator_address": "cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707"
    },
    {
      "reward": [{ "denom": "uatom", "amount": "55571.973980915812000000" }],
      "validator_address": "cosmosvaloper14lultfckehtszvzw4ehu0apvsr77afvyju5zzy"
    }
  ]
}
