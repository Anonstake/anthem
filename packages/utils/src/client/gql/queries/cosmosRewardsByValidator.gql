query cosmosRewardsByValidator($address: String!){
    cosmosRewardsByValidator(address: $address){
        reward{
            denom
            amount
        }
        validator_address
    }
}