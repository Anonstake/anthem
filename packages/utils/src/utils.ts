import bech32 from "bech32";
import { NETWORKS, NetworkDefinition, NETWORK_NAME } from "./networks";

/**
 * Assert a condition cannot occur. Used for writing exhaustive switch
 * blocks (e.g. see unwrapOkValueIfExists).
 */
export const assertUnreachable = (x: never): never => {
  throw new Error(
    `Panicked! Received a value which should not exist: ${JSON.stringify(x)}`,
  );
};

/**
 * Determine the network for a given address using the address prefix.
 */
export const deriveNetworkFromAddress = (
  address: string,
): NetworkDefinition => {
  if (address.substring(0, 6) === "cosmos") {
    return NETWORKS.COSMOS;
  } else if (address.substring(0, 5) === "terra") {
    return NETWORKS.TERRA;
  } else if (address.substring(0, 4) === "kava") {
    return NETWORKS.KAVA;
  } else if (address.length === 44) {
    return NETWORKS.OASIS;
  }

  throw new Error(
    `Unrecognized address ${address} with no associated network!`,
  );
};

/**
 * Get the network definition from a provided network name.
 */
export const getNetworkDefinitionFromIdentifier = (networkName: string) => {
  const name = networkName.toUpperCase();

  if (name in NETWORKS) {
    return NETWORKS[name];
  }

  throw new Error(
    `Unsupported or invalid network name ${networkName} supplied!`,
  );
};

/**
 * Get the network definition from a provided network price ticker.
 */
export const getNetworkDefinitionFromTicker = (ticker: string) => {
  const network = Object.values(NETWORKS).find(n => n.ticker === ticker);

  if (network) {
    return network;
  } else {
    throw new Error(
      `Unsupported or invalid network ticker ${ticker} supplied!`,
    );
  }
};

/**
 * Get network enums.
 */
const getCosmosSdkAddressEnumsForNetwork = (name: NETWORK_NAME) => {
  const networkPrefix = name.toLowerCase();
  return {
    ACCOUNT_ADDRESS: `${networkPrefix}`,
    ACCOUNT_PUBLIC_KEY: `${networkPrefix}pub`,
    VALIDATOR_OPERATOR_ADDRESS: `${networkPrefix}valoper`,
    VALIDATOR_CONSENSUS_ADDRESS: `${networkPrefix}valcons`,
    VALIDATOR_CONSENSUS_PUBLIC_KEY: `${networkPrefix}valconspub`,
    VALIDATOR_OPERATOR_PUBLIC_KEY: `${networkPrefix}valoperpub`,
  };
};

/**
 * Convert a validator address to its associated delegator address.
 */
export const validatorAddressToOperatorAddress = (validatorAddress: string) => {
  const decodedAddress = bech32.decode(validatorAddress);
  const operatorAddress = bech32.encode("cosmos", decodedAddress.words);
  return operatorAddress;
};

/**
 * Decode a validator address using bech32 and re-encode it to derive the
 * associated validator address.
 */
export const getValidatorAddressFromDelegatorAddress = (
  address: string,
  network: NETWORK_NAME,
): string | null => {
  try {
    const networkAddressEnum = getCosmosSdkAddressEnumsForNetwork(network);
    const decodedAddress = bech32.decode(address);
    const validatorAddress = bech32.encode(
      networkAddressEnum.VALIDATOR_OPERATOR_ADDRESS,
      decodedAddress.words,
    );

    return validatorAddress;
  } catch (err) {
    return null;
  }
};
