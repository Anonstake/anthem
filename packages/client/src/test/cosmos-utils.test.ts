import { NETWORKS } from "@anthem/utils";
import { transaction } from "test/data/mock-transaction-data.json";
import {
  createCosmosTransactionPostBody,
  createDelegationTransactionMessage,
  createRewardsClaimTransaction,
  createTransactionRequestMetadata,
} from "tools/cosmos-utils";
import { cosmosAccountInformation } from "../../../utils/src/client/data/cosmosAccountInformation.json";
import { cosmosRewardsByValidator } from "../../../utils/src/client/data/cosmosRewardsByValidator.json";

describe("cosmos-utils", () => {
  test("createDelegationTransactionMessage", () => {
    const result = createDelegationTransactionMessage({
      amount: "50",
      address: "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
      gasAmount: "1500",
      gasPrice: "150000",
      denom: "uatom",
      network: NETWORKS.COSMOS,
      validatorOperatorAddress:
        "cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707",
    });

    expect(result).toMatchSnapshot();
  });

  test.skip("createRewardsClaimTransaction", () => {
    /**
     * TODO: Fix this test!
     */
    const result = createRewardsClaimTransaction({
      address: "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
      gasAmount: "1500",
      gasPrice: "150000",
      denom: "uatom",
      // @ts-ignore
      selectedRewards: cosmosRewardsByValidator,
    });

    expect(result).toMatchSnapshot();
  });

  test("createTransactionRequestMetadata", () => {
    const result = createTransactionRequestMetadata({
      address: "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
      gasAmount: "1500",
      gasPrice: "150000",
      account: cosmosAccountInformation,
      network: NETWORKS.COSMOS,
    });

    expect(result).toMatchSnapshot();
  });

  test("createCosmosTransactionPostBody", () => {
    const { value } = transaction.tx;
    const { pub_key, signature } = value.signatures[0];
    value.signatures = [];
    const metadata = {
      account_number: "146",
      chain_id: "cosmoshub-3",
      fees: "225000000",
      from: "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
      generate_only: false,
      sequence: "44",
    };
    const transactionData = {
      txMsg: value,
      txRequestMetadata: metadata,
    };
    const result = createCosmosTransactionPostBody({
      transactionData,
      signature,
      publicKey: pub_key.value,
    });
    expect(result).toMatchSnapshot();
  });
});
