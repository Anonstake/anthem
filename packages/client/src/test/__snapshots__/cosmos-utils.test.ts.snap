// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cosmos-utils createCosmosTransactionPostBody 1`] = `
Object {
  "type": "cosmos-sdk/StdTx",
  "value": Object {
    "fee": Object {
      "amount": Array [
        Object {
          "amount": "930",
          "denom": "uatom",
        },
      ],
      "gas": "37188",
    },
    "memo": "Stake online with Chorus One at https://anthem.chorus.one",
    "msg": Array [
      Object {
        "type": "cosmos-sdk/MsgSend",
        "value": Object {
          "amounts": Array [
            Object {
              "amount": "2000000",
              "denom": "uatom",
            },
          ],
          "from_address": "cosmos15urq2dtp9qce4fyc85m6upwm9xul3049um7trd",
          "to_address": "cosmos16nte2qf5l0u39s86wcwu4fff4vdtvg7yn3uksu",
        },
      },
    ],
    "signatures": Array [
      Object {
        "account_number": "146",
        "pub_key": Object {
          "type": "tendermint/PubKeySecp256k1",
          "value": "AoUAnGnBsT7xGv2PTnIQbfK4t4oNyimGk2AhPwfgxq3T",
        },
        "sequence": "44",
        "signature": "11vxO02HrMMPwm3KfXjj2rBZeTxEEyMPS7TlK185f3J9O8dJJD30GgBd4SFyyoLi0KfW14Mh2uR5mPT5NHbT3w==",
      },
    ],
  },
}
`;

exports[`cosmos-utils createDelegationTransactionMessage 1`] = `
Object {
  "fee": Object {
    "amount": Array [
      Object {
        "amount": "225000000",
        "denom": "uatom",
      },
    ],
    "gas": "1500",
  },
  "memo": "Stake online with Chorus One at https://anthem.chorus.one",
  "msg": Array [
    Object {
      "type": "cosmos-sdk/MsgDelegate",
      "value": Object {
        "amount": Object {
          "amount": "50000000",
          "denom": "uatom",
        },
        "delegator_address": "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
        "validator_address": "cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707",
      },
    },
  ],
  "signatures": null,
}
`;

exports[`cosmos-utils createRewardsClaimTransaction 1`] = `
Object {
  "fee": Object {
    "amount": Array [
      Object {
        "amount": "225000000",
        "denom": "uatom",
      },
    ],
    "gas": "1500",
  },
  "memo": "Stake online with Chorus One at https://chorus.one",
  "msg": Array [
    Object {
      "type": "cosmos-sdk/MsgWithdrawDelegationReward",
      "value": Object {
        "delegator_address": "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
        "validator_address": "cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707",
      },
    },
    Object {
      "type": "cosmos-sdk/MsgWithdrawDelegationReward",
      "value": Object {
        "delegator_address": "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
        "validator_address": "cosmosvaloper14lultfckehtszvzw4ehu0apvsr77afvyju5zzy",
      },
    },
  ],
  "signatures": null,
}
`;

exports[`cosmos-utils createTransactionRequestMetadata 1`] = `
Object {
  "account_number": "16451",
  "chain_id": "cosmoshub-3",
  "fees": "225000000",
  "from": "cosmos1yeygh0y8rfyufdczhzytcl3pehsnxv9d3wsnlg",
  "generate_only": false,
  "sequence": "75",
}
`;
